### PR TITLE
Feat/prefill ids

### PR DESCRIPTION
**Pull request by valentindey**

### CODE REVIEW

File: app/scripts/importUiMetaData/README.md [hunk 1]
```diff
@@ -40,19 +40,21 @@ This field specifies the correct input component for the given json path.
...
+| STRING_PRODUCT_ID         | strings referencing a product ID<br/> allows selecting from matching product IDs in the frontend                                           |
+| WITH_GENERATED_GROUP_ID   | objects with a unique `group_id` prefilled                                                                                                 |
+| STRING_GROUP_ID           | strings referencing a group ID<br/> allows selecting from matching group IDs in the frontend                                               |
+| OBJECT_CWE                | special type for CWE objects ensuring that ID and name match                                                                               |
+| OBJECT_CVSS_2             | special CVSS v2 object                                                                                                                     |
+| OBJECT_CVSS_3             | special CVSS v3.0 / v3.1object                                                                                                             |
```
Comment (tschmidtb51) [inline app/scripts/importUiMetaData/README.md:56]:
```suggestion
| OBJECT_CVSS_3             | special CVSS v3.0 / v3.1 object                                                                                                             |

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 2]
```diff
@@ -384,4 +384,43 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+      cy.get(
+        `[data-testid="menu_entry-/product_tree/relationships-add_item_button"]`
+      ).click()
+      cy.get(
+        `[data-testid="menu_entry-/product_tree/relationships/0/full_product_name"]`
+      ).click()
+      cy.get(
+        '[data-testid="attribute-product_tree-relationships-0-full_product_name-product_id"] input'
+      ).should('have.value', 'CSAFPID-0002')
+    }
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:424]: I guess also a test should be added here that checks on the group Id...

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 3]
```diff
@@ -384,4 +384,43 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+      cy.get(
+        '[data-testid="menu_entry-/product_tree/branches-add_item_button"]'
+      ).click({ force: true })
+      cy.get(
+        '[data-testid="menu_entry-/product_tree/branches/0/product"]'
+      ).click()
+      cy.get(
+        '[data-testid="attribute-product_tree-branches-0-product-product_id"] input'
+      ).should('have.value', 'CSAFPID-0001')
+
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:412]: Add a test here for an item in `full_product_names` element.

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
The parent element is not visible and therefore the test fails.

However, it also fails a manual test, as the `product_id` is assigned to the first (most upper) `branch` automatically. Subsequent ones don't get one. It is correct that you need only one but it needs to be assigned at the leaf, not the root. Please make sure to add a test for such a nested structure.

**Comment by tschmidtb51** [comment]
I guess we need to discuss some details here, flagging @mfd2007

**Comment by valentindey** [comment]
> The parent element is not visible and therefore the test fails.

It was a little late yesterday and I forgot to commit a fix for that. Added now.

I'll get back at your other comments later.

**Comment by valentindey** [comment]
> However, it also fails a manual test, as the `product_id` is assigned to the first (most upper) `branch` automatically. Subsequent ones don't get one. It is correct that you need only one but it needs to be assigned at the leaf, not the root. Please make sure to add a test for such a nested structure.

This can not be realized (easily without additional rules) as we don't know at what point we're adding a leaf node. I changed the behavior to add a product ID for all branche items in the structure which is more consistent and also reflects the behavior of the old version.

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 2 above)
Comment (valentindey) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:456]: Added a test for prefilling group IDs

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 3 above)
Comment (valentindey) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:412]: extended the test

**Comment by valentindey** [comment]
I added a different uiType `STRING_GENERATE_PRODUCT_ID` (instead of `WITH_GENERATED_PRODUCT_ID` for the parent object) that will get an additional button (pencil symbol) that when clicked will fill the corresponding field with a unique product id.

For other types it should be easy to add such buttons for generating values in the future.

As discussed, the group ids are always prefilled.

File: app/lib/app/SecvisogramPage/View/FormEditor/editors/GenericEditor/Attributes/shared/Attribute.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import { faInfoCircle } from '@fortawesome/free-solid-svg-icons'
+import { faInfoCircle, faPencil } from '@fortawesome/free-solid-svg-icons'
```
Comment (tschmidtb51) [inline app/lib/app/SecvisogramPage/View/FormEditor/editors/GenericEditor/Attributes/shared/Attribute.js:1]: I think `faMagic` is a better fit.
```suggestion
import { faInfoCircle, faMagic } from '@fortawesome/free-solid-svg-icons'

File: app/lib/app/SecvisogramPage/View/FormEditor/editors/GenericEditor/Attributes/shared/Attribute.js [hunk 5]
```diff
@@ -93,6 +99,19 @@ export default function Attribute({
...
+          <button
+            title="generate value"
+            data-testid={instancePath.join('-') + '-generateButton'}
+            type="button"
+            className="w-6 h-6 flex-none text-slate-400 hover:text-slate-800 m-1"
+            onClick={() => {
+              updateDoc(instancePath, generateFn())
+            }}
+          >
+            <FontAwesomeIcon icon={faPencil} />
```
Comment (tschmidtb51) [inline app/lib/app/SecvisogramPage/View/FormEditor/editors/GenericEditor/Attributes/shared/Attribute.js:112]: Then use it here:
```suggestion
            <FontAwesomeIcon icon={faMagic} />

File: app/scripts/importUiMetaData/README.md [hunk 6]
```diff
@@ -40,19 +40,21 @@ This field specifies the correct input component for the given json path.
...
+| STRING_DATETIME           | datetime values                                                                                                                            |
+| STRING_URI                | strings matching the URI pattern                                                                                                           |
+| STRING_ENUM               | strings with predefined values to select from<br/> the values are defined in the schema file                                               |
+| STRING_WITH_OPTIONS       | strings with predefined values to select from, still allowing other user input<br/> the can be added via the `options` property, see below |
+| STRING_MULTI_LINE         | strings that are potentially larger                                                                                                        |
+| WITH_GENERATED_PRODUCT_ID | objects with a unique `product_id` prefilled                                                                                               |
```
Comment (tschmidtb51) [inline app/scripts/importUiMetaData/README.md:50]: I guess that one does not longer exist...

File: app/lib/app/SecvisogramPage/View/FormEditor/schema.js [hunk 7]
```diff
@@ -4521,6 +4524,7 @@ export default /** @type {const} */ ({
                             description: 'FullProductNameDescription',
                           },
                           addMenuItemsForChildObjects: true,
+                          uiType: 'WITH_GENERATED_PRODUCT_ID',
```
Comment (tschmidtb51) [inline app/lib/app/SecvisogramPage/View/FormEditor/schema.js:4527]: Hmm - I think that one does not longer exist...

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Please see my comments. Also: If you hover over the field, it shows the error message "Please enter a URL." as the type is set to `url` somehow/somewhen...

**Comment by tschmidtb51** [comment]
Screenshot from 2023-02-21 22-49-28
This picture shows the error as well as the updated icon.

File: app/lib/app/SecvisogramPage/View/FormEditor/editors/GenericEditor/Attributes/shared/Attribute.js (hunk 4 above)
Comment (valentindey) [reply app/lib/app/SecvisogramPage/View/FormEditor/editors/GenericEditor/Attributes/shared/Attribute.js]: That suits the function even better, thanks for the suggestion. Updated.

File: app/scripts/importUiMetaData/README.md (hunk 6 above)
Comment (valentindey) [reply app/scripts/importUiMetaData/README.md]: True, I updated the readme.

File: app/lib/app/SecvisogramPage/View/FormEditor/schema.js (hunk 7 above)
Comment (valentindey) [reply app/lib/app/SecvisogramPage/View/FormEditor/schema.js]: Yes, I updated it.

**Comment by valentindey** [comment]
Thanks for the thorough review @tschmidtb51 . I updated everything according to your suggestions, please have another look.

**Review by tschmidtb51 — APPROVED** [review]
LGTM

**Pull request merged**